File: morning_exercises/day26.cpp
#include <ios>
#include <vector>
#include <iostream>

std::vector<std::vector<int>> courseGraph(int number, std::vector<std::pair<int, int>>& prerequisites) {
    std::vector<std::vector<int>> graph(number);
    for (const auto& [a, b] : prerequisites) graph[b].push_back(a);
    return graph;
}

bool checkDfs(std::vector<std::vector<int>>& graph, int node, std::vector<int>& state) {
    if (state[node]==2) return true;
    if (state[node]==1) return false;

    state[node]=1;
    for (int neighbour : graph[node]) if (!checkDfs(graph, neighbour, state)) return false;
    state[node]=2;
    return true;
}

bool isDoable(std::vector<std::vector<int>>& graph) {
    std::vector<int> state(graph.size());
    for (int i = 0; i<graph.size(); i++) if (!checkDfs(graph, i, state)) return false;
    return true;
}

int main() {
    std::vector<std::pair<int,int>> prerequisites = {{1,2}, {2,3}};
    auto graph = courseGraph(4, prerequisites);

    std::cout << std::boolalpha;
    std::cout << isDoable(graph) << "\n";
}

File: morning_exercises/day26.py
def courseSchedule(numCourses: int, prerequisites: list):
    graph=[[] for _ in range(numCourses)]
    for a, b in prerequisites:
        graph[b].append(a)
    return graph

def dfsCheck(graph: list, node: int, state: list)->bool:
    if state[node]==2: return True
    if state[node]==1: return False

    state[node]=1
    for req in graph[node]:
        if not dfsCheck(graph, req, state): return False
    state[node]=2
    return True

def isDoable(graph):
    state = [0 for _ in range(len(graph))]
    for i in range(0, len(graph)):
        if not dfsCheck(graph, i, state): return False
    return True

courseschedule = courseSchedule(3, [[1,0], [2,1]])
print(isDoable(courseschedule))